src/OrleansServiceBus/Providers/Streams/EventHub/EventHubQueueCache.cs

```diff
@@ -133,38 +133,74 @@ public bool TryGetNextMessage(object cursorObj, out IBatchContainer message)
-            }
+            cachePressureMonitor.RecordCachePressureContribution(
+                TryCalculateCachePressureContribution(message.SequenceToken, out cachePressureContribution)
+                    ? cachePressureContribution
+                    : 0.0);
             return true;
         }
 
-        private class AveragingCachePressureMonitor
+    }
+
+    internal class AveragingCachePressureMonitor
+    {
+        private static readonly TimeSpan checkPeriod = TimeSpan.FromSeconds(2);
+        const double pressureThreshold = 1.0 / 3.0;
+        private readonly Logger logger;
+
+        private double accumulatedCachePressure;
+        private double cachePressureContributionCount;
+        private DateTime nextCheckedTime;
+        private bool isUnderPressure;
+
+        public AveragingCachePressureMonitor(Logger logger)
         {
-            const double pressureThreshold = 1.0/3.0;
+            this.logger = logger.GetSubLogger("-flowcontrol");
+            nextCheckedTime = DateTime.MinValue;
+            isUnderPressure = false;
+        }
 
-            private double accumulatedCachePressure;
-            private int cachePressureContributionCount;
+        public void RecordCachePressureContribution(double cachePressureContribution)
+        {
+            // Weight unhealthy contributions thrice as much as healthy ones.
+            // This is a crude compensation for the fact that healthy consumers wil consume more often than unhealthy ones.
+            double weight = cachePressureContribution < pressureThreshold ? 1.0 : 3.0;
+            accumulatedCachePressure += cachePressureContribution * weight;
+            cachePressureContributionCount += weight;
+        }
 
-            public void RecordCachePressureContribution(double cachePressureContribution)
+        public bool IsUnderPressure(DateTime utcNow)
+        {
+            if (nextCheckedTime < utcNow)
             {
-                accumulatedCachePressure += cachePressureContribution;
-                cachePressureContributionCount++;
+                CalculatePressure();
+                nextCheckedTime = utcNow + checkPeriod;
             }
+            return isUnderPressure;
+        }
 
-            public bool IsUnderPressure()
+        private void CalculatePressure()
+        {
+            // if we don't have any contributions, don't change status
+            if (cachePressureContributionCount < 0.5)
             {
-                if (cachePressureContributionCount == 0)
-                    return false;
-
-                double pressure = accumulatedCachePressure/cachePressureContributionCount;
-
-                cachePressureContributionCount = 0;
-                accumulatedCachePressure = 0;
+                // after 5 checks with no contributions, check anyway
+                cachePressureContributionCount += 0.1;
+                return;
+            }
```

Comment:
does this mean that if there if pressure is high and the consumers are not consuming any events because they are stalled, we will re-start reading every 10 seconds?